Pullquote: add block support

ramonjd opened the pull request:
## Description

In this PR, we're adding border and color block support to the Pullquote block. 

Given that we can now achieve border and background variations using block supports, we're removing the existing styles and adding a `textAlign` attribute to compensate.

Furthermore, because the new block supports already offers the controls we need to style the Pullquote, the current controls are not required. We have therefore removed those and also the styles/border/classes logic.

ℹ️ There was [some discussion](https://github.com/WordPress/gutenberg/pull/30834#issuecomment-819951425) about whether this PR could be split into two: adding first border support, then color support in a later PR. I believe this block is a special case because, at present, the background colour control determines **both** the border colour and background colour, depending on which block style we select.

## How has this been tested?

Manually, in the editor (web and mobile device)

Because this PR deprecates the current version of Pullquote, we've updated the fixtures (version 4 is the last previous version).

Check that fixtures are valid:

`npm run fixtures:regenerate test/integration/full-content/full-content.test.js`

And by pasting current fixture block code into the editor and making sure a valid transform takes place.

Check in a native mobile device that the block works as intended.

## Screenshots <!-- if applicable -->
<img width="1046" alt="Screen Shot 2021-04-27 at 4 33 11 pm" src="https://user-images.githubusercontent.com/6458278/116196024-4fb25980-a776-11eb-9c33-5881adbf0354.png">


## Types of changes
These changes add experimental border and color block support to the Pullquote block.

The attributes are changing, hence we've added a deprecation (v4).

Block support controls supersede the block's original controls, so we have removed them. So too, there's now no longer any need to differentiate between the usage of `mainColor` for background and border, since we can control each independently.

Part of https://github.com/WordPress/gutenberg/issues/28913

cc @aaronrobertshaw

ramonjd committed:
initial commit

ramonjd committed:
Updating fixtures and deprecated fixtures.
Adding some TODOs for later cleanups

ramonjd committed:
Removed block attributes from block.json that block supports injects

ockham commented on packages/e2e-tests/fixtures/blocks/core__pullquote__deprecated-4.parsed.json:
```diff
@@ -0,0 +1,24 @@
+            "\n<figure class=\"wp-block-pullquote has-background has-black-background-color is-style-solid-color\"><blockquote class=\"has-text-color has-white-color\"><p>pullquote</p><cite>before block supports</cite></blockquote></figure>\n"
+        ]
+    },
+    {
+        "blockName": null,
+        "attrs": {},
+        "innerBlocks": [],
+        "innerHTML": "\n",
+        "innerContent": [
+            "\n"
+        ]
+    }
```
This is a bit weird -- is the parser tripping over something? 🤔

ramonjd replied:
Thanks @ockham 

I _think_ I've seen this before. When a new line exists at the end of a html fixture file, it seems as if the parser interprets this as an empty block. 😱 

The result is an empty block object in the parsed JSON (`*.parsed.json` files).


```
   {
        "blockName": null,
        "attrs": {},
        "innerBlocks": [],
        "innerHTML": "\n",
        "innerContent": [
            "\n"
        ]
    }
```

Changing [this function](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/e2e-tests/fixtures/utils.js#L61) to something like 

```
function getBlockFixtureHTML( basename ) {
	const filename = `${ basename }.html`;
	const fileContents = readFixtureFile( FIXTURES_DIR, filename );
	return {
		filename,
		file: fileContents ? fileContents.trim() : null,
	};
}
```

 might help?

I'll try it out later.

ramonjd replied:
I created a PR to remove them: https://github.com/WordPress/gutenberg/pull/30981

aaronrobertshaw replied:
From my experience the empty block object is caused by a blank line at the end of the fixture file. Thanks for putting together the PR to clean up all fixtures 👍 

I think you'll also need to update the new pullquote deprecation in this PR, removing the blank line from it. The original deprecations are caught in #30981 by the looks of it.

ockham commented on packages/e2e-tests/fixtures/blocks/core__pullquote__main-color.json:
```diff
@@ -1,16 +1,21 @@
+        "isValid": true,
+        "attributes": {
+            "value": "<p>Pullquote custom color</p>",
+            "citation": "before block supports",
```
The citation text is a bit misleading here, no? Since this isn't really a deprecation fixture, but a "regular" one.

Also, should we maybe rename this set of tests to `core__pullquote__color.*` (or `text-color`, `background-color`,...?) -- since the attribute isn't called `mainColor` any more?

ramonjd replied:
That was a copy 🍝 oversight. 👍

ramonjd replied:
> Also, should we maybe rename this set of tests to core__pullquote__color.* (or text-color, background-color,...?) -- since the attribute isn't called mainColor any more?

Good idea, thanks. 

I went with `*__custom-colors__* to test the new style attribute against the deprecation of `mainColor` & `customMainColor`, which (I think) is more in the spirit of the [original fixture](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/fixtures/blocks/core__pullquote__main-color.html).

ockham commented on packages/e2e-tests/fixtures/blocks/core__pullquote__main-color.parsed.json:
```diff
@@ -1,22 +1,29 @@
+            "\n<figure class=\"wp-block-pullquote is-style-default has-border-color has-white-color has-blue-background-color has-text-color has-background\" style=\"border-color:#2207d0\"><blockquote><p>Pullquote custom color</p><cite>before block supports</cite></blockquote></figure>\n"
+        ]
+    },
+    {
+        "blockName": null,
+        "attrs": {},
+        "innerBlocks": [],
+        "innerHTML": "\n",
+        "innerContent": [
+            "\n"
+        ]
+    }
```
Ah, another one of [those](https://github.com/WordPress/gutenberg/pull/30951/files?diff=split&w=1#r615777377) -- this time pre-existing 🤔

ramonjd replied:
Maybe the same as [above](https://github.com/WordPress/gutenberg/pull/30951#pullrequestreview-638772735). I'll test my hunch later.

ockham commented:
This might not be specific to this block, but if I set a border color, width, and radius, and leave the border style as 'Default', I see something like this:

![image](https://user-images.githubusercontent.com/96308/115237151-81934280-a11c-11eb-84ea-d09852352e47.png)

That looks a bit off. If I change the border style to 'Solid', it looks right:

![image](https://user-images.githubusercontent.com/96308/115237233-a12a6b00-a11c-11eb-8f70-8cabd9526b7b.png)

ockham commented on packages/block-library/src/pullquote/deprecated.js:
```diff
@@ -57,7 +58,128 @@ function parseBorderColor( styleString ) {
+				'color',
+				textColor
+			);
+			const blockquoteClasses =
+				( textColor || customTextColor ) &&
+				classnames( 'has-text-color', {
+					[ blockquoteTextColorClass ]: blockquoteTextColorClass,
+				} );
```
It seems that:

1. If either `textColor` or `customTextColor` is set, we want to set the `has-text-color` className. That figures 👍 
2. If one of those is set, _and_ `blockquoteTextColorClass` is set, we want to set the `[ blockquoteTextColorClass ]` class name. But IIUC, we only get a `blockquoteTextColorClass` if we have a non-custom text color (see [implementation](https://github.com/WordPress/gutenberg/blob/fe379134644e6a8b43a220d1d6e956def9b5295e/packages/block-editor/src/components/colors/utils.js#L49-L64)), so the `customTextColor` criterion is somewhat deceptive.

So can we maybe change this to

```suggestion
			const blockquoteClasses = classnames( {
				'has-text-color': textColor || customTextColor,
				[ blockquoteTextColorClass ]: blockquoteTextColorClass,
			} );
```

for better legibility, and to better match [`figureClasses`](https://github.com/WordPress/gutenberg/pull/30951/files#diff-9f8162c8a4e64eafe65af6866d3d2ffb30cc8dc72b71e5a13ab9f783865b038eR89-R92)?

ramonjd replied:
The above is much better, thanks.

I left everything in the save function "as is", since I wasn't sure whether to touch the deprecated save function, or to preserve as it is for succeeding archeological expeditions 😄

ockham commented on packages/block-library/src/pullquote/deprecated.js:
```diff
@@ -57,7 +58,128 @@ function parseBorderColor( styleString ) {
+			const isSolidColorStyle = includes( className, SOLID_COLOR_CLASS );
+
+			const style = {};
+			// Block supports: Set style.border.color if a deprecated block has a default style and a `customMainColor` attribute.
+			if ( ! isSolidColorStyle && customMainColor ) {
+				style.border = {
+					color: customMainColor,
+				};
+			}
+
+			// Block supports: Set style.color.background if a deprecated block has a solid style and a `customMainColor` attribute.
+			if ( isSolidColorStyle && customMainColor ) {
+				style.color = {
+					background: customMainColor,
+				};
+			}
```
Maybe turn this into a nested

```suggestion
			if ( customMainColor ) {
				if ( ! isSolidColorStyle ) {
					// Block supports: Set style.border.color if a deprecated block has a default style and a `customMainColor` attribute.
					style.border = {
						color: customMainColor,
					};
				} else {
					// Block supports: Set style.color.background if a deprecated block has a solid style and a `customMainColor` attribute.
					style.color = {
						background: customMainColor,
					};
				}
			}
```

to make the logical branching more explicit?

ockham commented on packages/e2e-tests/fixtures/blocks/core__pullquote__deprecated-3.serialized.html:
```diff
@@ -1,3 +1,3 @@
-<!-- wp:pullquote {"customMainColor":"#cd2653","textColor":"secondary"} -->
-<figure class="wp-block-pullquote" style="border-color:#cd2653"><blockquote class="has-text-color has-secondary-color"><p>pullquote</p></blockquote></figure>
+<!-- wp:pullquote {"textColor":"secondary"} -->
+<figure class="wp-block-pullquote has-secondary-color has-text-color"><blockquote><p>pullquote</p></blockquote></figure>
```
⚠️ Seems like we're losing the `customMainColor` attribute and style here?

ramonjd replied:
Hmm, I'm not sure why it's stripping the inline style. I think I need to relook at the v4 migration function. Thanks for spotting that one!

ramonjd replied:
Actually it was the v3 migration function. I'm not 100% certain if what I've done is kosher, but it seems that we need to add the block supports `style` attribute to the migrated attributes so that the inline style is added.

I guess the `customMainColor` attribute is stripped because no longer a valid attribute 🤞

aaronrobertshaw replied:
I might be a bit late to the conversation and missing something so feel free to ignore this if it doesn't apply or help.

When a block's attributes have changed and you need to add a deprecation with a `migrate` function. All the deprecations need to handle migrating the attributes correctly to the block's latest state or attribute definition, if that makes sense.

So if we have a deprecated attribute in a deprecation's attribute definitions, that would need migrating to the latest version of the block's attributes via the deprecation's `migrate` function.

In other words, each deprecation is responsible for bringing that version of the block up all the way up to the latest.

ramonjd replied:
> When a block's attributes have changed and you need to add a deprecation with a migrate function. All the deprecations need to handle migrating the attributes correctly to the block's latest state or attribute definition, if that makes sense.

I had suppressed an inkling this might be the case. :) Thanks @aaronrobertshaw, I'll have a closer look.

ockham commented on packages/e2e-tests/fixtures/blocks/core__pullquote__main-color__deprecated-4.serialized.html:
```diff
@@ -0,0 +1,3 @@
+<!-- wp:pullquote {"className":"has-background is-style-solid-color","style":{"color":{"background":"#48ba90","text":"#1567eb"}}} -->
+<figure class="wp-block-pullquote has-background is-style-solid-color has-text-color has-background" style="background-color:#48ba90;color:#1567eb"><blockquote><p>pullquote</p><cite>before block supports</cite></blockquote></figure>
```
Seems like the (text) `color` style is moved from `figure` to `blockquote` during the deprecation:

[Before](https://github.com/WordPress/gutenberg/pull/30951/files#diff-506a54e9a0743b84f3ef2a8ed507e4dfea5186df0e9776386d1852e75374d347):

```html
<figure style="background-color:...">
	<blockquote style="color:...">
		...
	</blockquote>
</figure>
```

After deprecation:

```html
<figure style="background-color:...; color:...">
	<blockquote>
		...
	</blockquote>
</figure>
```

I'm not sure about the impact, but any chance we can keep the HTML unchanged? I think we should be able to use our new `__experimentalSkipSerialization` mechanism to, well skip the "automagic" saving of the `color` attribute, and save it manually instead. cc/ @nosolosw 

(Not really a concern for the deprecation so much as for the block's `save()` method I think, but it's most easily visible here.)

ramonjd replied:
> I'm not sure about the impact, but any chance we can keep the HTML unchanged? I think we should be able to use our new __experimentalSkipSerialization mechanism to, well skip the "automagic" saving of the color attribute

So if we use `__experimentalSkipSerialization: true` in `supports.color`, I'm guessing we'd have to do all the work [gutenberg_apply_colors_support()](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/colors.php#L66) does, and therefore put back the corresponding color CSS classes (e.g., `has-background-color has-%s-background-color`), and text color style to the blockquote  etc.

Thanks for the sanity check! I'm still getting my head around things. 

> Seems like the (text) color style is moved from figure to blockquote during the deprecation:

I think I agree about re-adding text color style to the block quote, if it avoids specificity/design issues, but I'm not sure why else we'd want to do it?

On a more general level, would we ever accept some level of mutation when "upgrading" the block to use block supports? Especially if the block supports mechanism does not impact the UI in any way when it adds classes/inline styles to the containing block?

aaronrobertshaw replied:
I believe the normal "ideal scenario" for using the block supports is to have those styles applied to the block's wrapper if at all possible.

In this case there are a few of benefits for applying the block support styles on the block's wrapper element:

- The block's code is a lot simpler, easier to grok and maintain
- Avoids duplicating the generation and application of styles
- Keeps the `__experimentalSkipSerialization` option as a last resort when style must be applied to inner elements.

As a deprecation is already required to migrate the old attributes to the new, it covers the slight change to the HTML avoiding block validation errors. Any CSS specificity concerns, I think would be rather minor, more toward the application of the font color being less specific on the `blockquote` than before.

Overall, my vote would be to continue with the current approach as opposed to using `__experimentalSkipSerialization`.

ockham replied:
Okay, sounds good. I can't seem to come up with a compelling scenario where moving the text color style would present a huge problem, so let's stick with what we have 👍

oandregal replied:
So, we're changing both the classes and the inline styles for the `color` property from `blockquote` to `figure`. Braindump some use cases in which this could break things:

- is there any element in `figure` that will take color with this change but previously didn't?
- are themes targeting `blockquote.has-color` to do anything (change would be problematic)? Or are they using `figure .has-color` (change would be fine)?


In general, we should be open to changes if it's for the better. We should also check that existing themes don't break. A good litmus tests are the default themes. If we aren't sure about what's the status: either refactor or do behavior changes, not both at the same time.

ramonjd replied:
Thanks for the feedback!

> If we aren't sure about what's the status: either refactor or do behavior changes, not both at the same time.

Just checking, by "refactor", do you mean using `__experimentalSkipSerialization` and reattaching the styles to the blockquote?

The behaviour changes, assuming you're referring to the existing background/border logic, must be addressed regardless if we are to update this block to use block supports. So if we do the above refactor we'll have to do both.

It's no problem to refactor, I just wanted to flag it.

I wish there were an MVP path. Ripping off the mixed behaviour band-aid now nevertheless will help folks down the line in my opinion.

It would save a lot of time and energy if we could remove Pullquote's exsiting and inconvenient background/border logic _and_ offer block supports at the same time (which we'd have to do anyway).

> We should also check that existing themes don't break. A good litmus tests are the default themes.

Thanks for this helpful tip.

Despite the direction we take from here, I'll still spend some time testing as many themes as I can.

Given the uncertainty surrounding the consequences of these changes and tendency for cautious conservatism, I think refactoring to accommodate existing HTML structures will ensure there's one less thing to test/worry about. 👍 Happy to take that road, it'll just take a bit more time.

aaronrobertshaw replied:
I've tested this PR with a range of themes (more theme testing would be beneficial though). 

The majority worked OK as is. 

There were a couple of issues encountered:
- A few themes don't generate or set `has-<color>-border-color` CSS classes so border colors aren't displayed on the frontend. The editor forces the color to an inline style though.
- A couple of themes just needed a `color: inherit` style in the block styles to display correctly
- The TwentyNineteen themes force white text color if the blockquote element itself doesn't have the `has-text-color` class.

| Good | Need `color: inherit` in block styles | Need `has-text-color` on blockquote |
|---|---|---|
| <ul><li>TT1 Blocks</li> <li>TwentyTwentyOne</li> <li>TwentyTwenty</li> <li>TwentyTwenty Blocks</li> <li>Armando</li> <li>Hever</li><li>Q</li><li>Seedlet</li></ul> | <ul><li>Ambitious</li><li>Maywood</li></ul> | <ul><li>Twenty Nineteen</li><li>Twenty Nineteen Blocks</li></ul> |

Themes that don't set the `has-<color>-border-color` styles:
- Ambitious
- Maywood
- TwentyNineteen
- TwentyTwenty

ramonjd replied:
Another concern I see is the potential confusion surround the styles.

The "Solid color" doesn't really mean much if you can set (and clear) a custom background. 

Switching between the two is ineffectual (aside from the alignment).

![pullquote-solid-colour](https://user-images.githubusercontent.com/6458278/116028726-296bbb80-a69b-11eb-8dfb-680a2476e59e.gif)

ockham replied:
Indeed 🤔  Should we consider dropping this UI control, and migrating the attribute to background color (plus maybe a text alignment attribute)?

ramonjd replied:
> Should we consider dropping this UI control, and migrating the attribute to background color (plus maybe a text alignment attribute)?

Thanks for the input. 

I agree. 

I'm looking into this right now. It will add to this already large PR, but I don't really see much of an avenue, at least for the future of this PR 🤣 

We can already revert any commits after we've tested them and find that it's over the top.

aaronrobertshaw replied:
It sounds like a natural evolution of this PR to me. I don't have strong opinions on splitting it into a second PR.

One benefit of adding a `textAlign` attribute and removing the block style now would be keeping the deprecations to a minimum. All the changes driving the need for the deprecation to map the style to attributes would be together and in context.

ramonjd replied:
I've removed the styles in https://github.com/WordPress/gutenberg/pull/30951/commits/835caac25a29722063c32549acba011f6045b2ea

I've also added a `textAlign` attribute.

Many themes do target the `.is-style-solid-color` class, however the class name persists in `className` for existing blocks. 

We're not removing any of the existing styles aside from those dealing with text alignment, which the new attribute now takes care of. 

Thanks to everyone for the input. 

Once again, I'm happy to split up this PR if folks things there is a more gentle approach.

ramonjd committed:
Renaming and updating deprecation v4 colors to reflect the usage of custom colors in the deprecation and newer versions.
Removing hanging empty blocks in parsed JSON

ramonjd committed:
Implementing code change suggestions in deprecated.js
save: to make blockquote class assignment more consistent with figure class assignment
migrate: to make customMainColor/style logic neater and clearer

Remove state wrapper around pullquote export
Rebuilding fixtures with the (for now) expected hanging block in parsed.json files. We could trim them in `getBlockFixtureHTML`

ramonjd committed:
This commit ensures that we're setting the style attribute required for deprecation v3 where there is no solid style, a mainColor attribute and an existing inline border style on the figure.
Regnerated fixtures.

Removing textColor from block attributes as they're adding by supports hooks

ramonjd commented:
> This might not be specific to this block, but if I set a border color, width, and radius, and leave the border style as 'Default', I see something like this:

It's the style variant CSS driving this I think. It's my opinion that we can remove the style variations in the next iteration as mention in the description.

aaronrobertshaw commented on packages/block-library/src/pullquote/block.json:
```diff
@@ -17,22 +17,24 @@
+			"background": true,
+			"link": true
+		},
+		"__experimentalBorder": {
+			"color": true,
+			"radius": true,
+			"style": true,
+			"width": true
+		}
```
We could simplify this config and simply make it `"__experimentalBorder": true`.

I don't see many new border support features being added on top of what's already there so it should be safe to do.

ramonjd replied:
Ah interesting. 

Out of curiosity, the properties don't matter then for these [specific checks](https://github.com/WordPress/gutenberg/blob/933c45a70808b590ca7c7a44cf4a73aafe53d35e/lib/block-supports/border.php#L17), e.g., `gutenberg_block_has_support( $block_type, array( '__experimentalBorder', 'radius' ) )`?

ockham replied:
Hmm, seems like they do matter though? 🤔 

https://github.com/WordPress/gutenberg/blob/933c45a70808b590ca7c7a44cf4a73aafe53d35e/lib/blocks.php#L388-L395

ramonjd replied:
> Hmm, seems like they do matter though? 🤔

That's what I can't work out 🤣 

I tested with `"__experimentalBorder": true` in pullquote/block.json and things were working, which is nice. 

But I couldn't see how they were working given the specific checks in blocks/border.php. 🤷

ramonjd replied:
Regardless, personally it has helped me having the support properties printed there explicitly, only as a convenient reminder of what the block supports :D

aaronrobertshaw replied:
> Hmm, seems like they do matter though?

This was an oversight of mine [when I refactored the border support](https://github.com/WordPress/gutenberg/pull/30124/commits/d87a2999b0cac30d479628b24ed4b9169a40fd93) to use the suggested `gutenberg_block_has_support`. Given this function doesn't check if the parent property is true it created a disjoint between support for static blocks and dynamic blocks.

I'm not certain if all block supports will wan't to allow the blanket opt-in so in place of updating `gutenberg_block_has_support` I've created a new PR to fix up the server side block support checks for border support, https://github.com/WordPress/gutenberg/pull/31020.

> That's what I can't work out 🤣
>
> I tested with "__experimentalBorder": true in pullquote/block.json and things were working, which is nice.

It "works" in this case because the only block support checks for the pullquote block are done via JS and the editor. The resulting CSS classes and styles from the block support are saved into the block's post content. This content is simply spat out on the frontend as is. 

A dynamic block such as the `core/post-tags` block is rendered via PHP. In this situation, the CSS classes and styles are retrieved via the code referenced in `blocks/border.php`

> Regardless, personally it has helped me having the support properties printed there explicitly, only as a convenient reminder of what the block supports :D

There isn't any real downside in being explicit here regarding which support is being opted into. So feel free to leave as is. 

If a block support feature such as `__experimentalBorder` is opted into via a boolean true value, it would mean that it would automatically adopt any new sub-features in the future. That could be a double edged sword depending on the scenario. Regarding borders, I don't think much will change there.

I'm glad the discussion came up though as it highlighted the disjoint between the editor and server side checks for support. Thanks! 👍

aaronrobertshaw commented:
> > This might not be specific to this block, but if I set a border color, width, and radius, and leave the border style as 'Default', I see something like this:
> 
> It's the style variant CSS driving this I think. It's my opinion that we can remove the style variations in the next iteration as mention in the description.

I believe this is a theme issue. The theme sets a solid border style for the top and bottom borders. The "Solid color" block style has [more specific CSS](https://github.com/ramonjd/gutenberg/blob/add/pullquote-block-border-color-support/packages/block-library/src/pullquote/style.scss#L34) that sets `border: none`.

<img width="313" alt="Screen Shot 2021-04-20 at 4 39 38 pm" src="https://user-images.githubusercontent.com/60436221/115352904-3b60e080-a1fb-11eb-9492-440e948f0d02.png">

aaronrobertshaw commented on packages/block-library/src/pullquote/block.json:
```diff
@@ -17,22 +17,24 @@
-		"customMainColor": {
-			"type": "string"
-		},
 		"textColor": {
```
We should be able to remove the `textColor` attribute as well.

ramonjd committed:
Regenerating fixtures after the trim fix in https://github.com/WordPress/gutenberg/pull/30981

ramonjd committed:
Ensure that child elements inherit their colors from the figure element

ramonjd committed:
All deprecation migration functions, where their corresponding save function require it, are returning the required attributes for the latest version.

ockham commented on packages/block-library/src/pullquote/deprecated.js:
```diff
@@ -152,24 +273,60 @@ const deprecated = [
-				mainColor,
+				backgroundColor: isSolidColorStyle ? mainColor : undefined,
+				borderColor: isSolidColorStyle ? undefined : mainColor,
+				style,
```
This will currently add `style`, even if it's an empty object, see https://github.com/WordPress/gutenberg/pull/30951/files#diff-cc59bbe727811591bfc343d05c2f46e36e9562c3e01319de8044160d437a8db5R1. We could consider omitting it in that case.

ramonjd replied:
Nice, thanks for catching that one 😄

ockham commented on packages/block-library/src/pullquote/deprecated.js:
```diff
@@ -252,6 +409,46 @@ const deprecated = [
+				className,
+				backgroundColor: isSolidColorStyle ? mainColor : undefined,
+				borderColor: isSolidColorStyle ? undefined : mainColor,
+				style,
```
Same as https://github.com/WordPress/gutenberg/pull/30951/files#r620242082

ockham commented on packages/block-library/src/pullquote/deprecated.js:
```diff
@@ -57,7 +58,127 @@ function parseBorderColor( styleString ) {
+				className,
+				backgroundColor: isSolidColorStyle ? mainColor : undefined,
+				borderColor: isSolidColorStyle ? undefined : mainColor,
+				style,
```
Same as https://github.com/WordPress/gutenberg/pull/30951/files#r620242082

ramonjd committed:
Since we've introduced block support for border and background color, the default and solid block styles were, aside from the text alignment, redundant. So we have removed them.
To compensate for the alignment, this commit adds a text alignment block control.
Deprecated blocks will still have the `is-style-solid-class`, so we keep the styles targeting it, and also ensure that we set the new attribute `textAlign` to "left" when it's present.

Add specificity to the left and right text align styles so that the block alignment controls have a great chance of overriding theme styles.

ockham commented:
Text alignment works for me on the frontend, but doesn't seem to work in the editor 😕  (This is with the TT1 Blocks theme.)

Here's the relevant part of the styling, per my browser inspector. Looks like a specificity issue (with `.wp-block-pullquote` getting the `.editor-styles-wrapper` added automaticcally).

![image](https://user-images.githubusercontent.com/96308/116396466-bb9fcb00-a825-11eb-9ea6-939d6e1fa8e2.png)

ockham commented on packages/block-library/src/pullquote/style.scss:
```diff
@@ -1,8 +1,16 @@
 	margin: 0 0 1em 0;
 	padding: 3em 0;
-	text-align: center;
+	text-align: center; // Default text-alignment is left.
```
Shouldn't that comment read

```suggestion
	text-align: center; // Default text-alignment is center.
```

?

ockham replied:
Ohh, I think I misread that. The comment expresses what the default alignment would be _without_ this setting, right?

ramonjd replied:
This comment is confusing, you're right. I think it was a personal note that I left in there while testing.

I think it should be something like: 

```
// Default text-alignment where the `textAlign` attribute value isn't specified.
```

ockham commented on packages/block-library/src/pullquote/edit.js:
```diff
@@ -129,9 +65,9 @@ function PullQuoteEdit( {
 							__( 'Add quote' )
 						}
-						textAlign="center"
+						textAlign="right"
```
How come we're changing this to `right`? 🤔

ramonjd replied:
Bad commit. I was testing something. Thanks for spotting. 👍

ockham commented on packages/block-library/src/pullquote/edit.js:
```diff
@@ -147,7 +83,7 @@ function PullQuoteEdit( {
 							className="wp-block-pullquote__citation"
 							__unstableMobileNoFocusOnMount
-							textAlign="center"
+							textAlign="right"
```
Same as https://github.com/WordPress/gutenberg/pull/30951/files#r622309663

ramonjd commented:
> Here's the relevant part of the styling, per my browser inspector. Looks like a specificity issue (with .wp-block-pullquote getting the .editor-styles-wrapper added automaticcally).

Thanks for testing again @ockham 

I think I wasn't running `npm run dev` or something 🤣 

Previously I tried `.wp-block-pullquote.has-text-align-left blockquote`, which appears to have the level of specificity required.

I then removed the `blockquote` selector to retest, and it was working, but clearly I was looking at the old build. 🤕 

Do folks see any problem with targeting the inner wrapper (e.g., the `blockquote`)?

`.wp-block-pullquote.has-text-align-left blockquote` 

I was going to give the `<blockquote />` a class name, e.g., 

`.wp-block-pullquote.has-text-align-left .wp-block-pullquote__blockquote`

but thought that semantics requires the use of `<blockquote />` anyway, therefore I wouldn't expect the element to change.

ockham commented:
Targetting the `<blockquote />` seems fine to me. As an alternative, I was considering suggesting even attaching the `has-text-align-*` class to that element; but I don't feel strongly about that, and since we're attaching all classes and styles to the `<figure />` now, it makes sense to be consistent there, I think.

ockham reviewed (approved):
Thanks @ramonjd, this is looking good now! 🚢 

Some e2e tests are still failing -- I believe they're unrelated. Maybe try a rebase?

ramonjd commented:
Hi @jasmussen 

Just wanted to make sure I ran this change by you as well, as it is quite a big change to the block. Thanks!

jasmussen commented:
Taking a look now.

jasmussen commented:
Thank you for the ping. Yes, this is a big one, this is what I see:

<img width="313" alt="Screenshot 2021-05-20 at 13 16 18" src="https://user-images.githubusercontent.com/1204802/118970010-0bc60500-b96e-11eb-826c-ee4c39e272ef.png">

To be blunt, that UI is not good enough at all. It takes up a great deal of space, it's a very far cry from what we need it to be as outlined in #27331, and it brings enormous prominence to design tools that they honestly don't deserve. There are color controls both in the Border panel, and in a dedicated separate Color panel. There are 5 different "clear" buttons, and the word "Border" is duplicated as many times.

I'm aware that #27331 with its progressive discovery and refactor of panels in the sidebar is a mouthful. But even smaller steps like #31333 or #31337 would make a big difference. I understand the desire to bring these features to bear, so that we can build more expressive patterns. At the same time, we've known for a long time that we have to improve these interfaces in the future. Which means that every single inspector we build using existing tools, we will have to come back and rewrite. 

Given that this PR is functional, has a green check, passes the tests, and adds the controls to the relatively exotic Pullquote block, is it okay to land it? Probably. But I would suggest we don't expand these controls beyond this block until we can make progress on #31337, #31333, or ideally #27331. I appreciate your hard work on this, and I understand the motivation. But in order to scale design tools, we have to make progress on the UI soon.

ramonjd commented:
Thanks for your quick and considered reply @jasmussen 

As you've hinted at, this one was always going to be a bit iffy since the blocks supports make the existing styles and attribute juggling partially redundant; hence the number of changes required to make deprecations work and so on.

I tried to find ways to split this one up into smaller iterations but the existing logic threw 🍌  peels in my path around most corners.

> But I would suggest we don't expand these controls beyond this block until we can make progress on #31337, #31333, or ideally #27331. 

I haven't detected an overwhelming urgency to migrate the controls for this particular block, so the option exists to wait until the UI controls lend themselves to a more appealing aesthetic and UX. 

In my opinion, the level of effort to migrate  and refactor this block will be similar independent of the controls UI due the reasons mentioned above. The work completed so far therefore won't have been done in vain. 

Still, because this bulk of this PR rewires the wily and 'exotic' pullquote block in anticipation of block supports, it might be worth, as you say, limiting further additions to this block, and treating it as a special case.

I'm happy to follow your and other folks' further advice on the matter.

aaronrobertshaw commented:
@jasmussen makes some great points regarding the UI. 

> But even smaller steps like #31333 or #31337 would make a big difference

Steps towards refining the border support UI are already being made in https://github.com/WordPress/gutenberg/pull/31585. 

It is still currently only in draft status, as to achieve the design, changes to the underlying block support mechanism are required. There are two PRs (https://github.com/WordPress/gutenberg/pull/31483, https://github.com/WordPress/gutenberg/pull/31641)  making these changes however they are still pending review. Any help getting those reviewed would help us move the UI forward 🙏 

> I'm aware that #27331 with its progressive discovery and refactor of panels in the sidebar is a mouthful.

The designs in #27331 are exciting however my understanding is a lot of that is based on G2 components or vice versa. There is work underway to begin the process of bringing the underlying systems and mechanisms required from G2 to implement the controls and panels in #27331 into Gutenberg. I don't believe it is something that can just be done piece by piece through PRs such as this.

It feels as though to make implementation of #27331 designs a requirement for PRs adding any form of control to the sidebar, effectively blocks any such PR until that large project is completed.

I appreciate we have a fine line to walk in terms of exposing controls & panels versus providing the functionality required by users, block patterns etc. I believe it was to assist in being able to land block support related changes that the UI for new block supports was disabled by default. Since the previous block styles did adjust borders it does appear the pullquote block would require the UI enabled though.

If we can get movement on the refinement of the border support UI, would you be happy if this was merged whenever those refinements are?

jasmussen commented:
Thank you for your thoughtful replies. 

_I had just written 20 minutes worth of comment here, and then I lost power. This will be my second attempt, so it will likely be a reduced version of it._ 😅 

High level, I think the question we have to ask ourselves when we build these features is: what designs do the new tools enable, and are they worth it? While [this is clearly just a basic example](https://user-images.githubusercontent.com/6458278/116196024-4fb25980-a776-11eb-9c33-5881adbf0354.png), it doesn't convey why we need expansive border tools urgently. It would be helpful with a pattern design to show what they can accomplish.

I had some comments around how to move forward with this PR: I think it boils down to how much work we'll need to redo when we get a chance to revisit this. If it requires deprecations, perhaps best to wait. If not, it might be fine? I'd still keep it Pullquote only for now, especially pending the separate PRs you have to address Jay's comments. Thank you for those!

---

> The designs in #27331 are exciting however my understanding is a lot of that is based on G2 components or vice versa. There is work underway to begin the process of bringing the underlying systems and mechanisms required from G2 to implement the controls and panels in #27331 into Gutenberg. I don't believe it is something that can just be done piece by piece through PRs such as this.

> It feels as though to make implementation of #27331 designs a requirement for PRs adding any form of control to the sidebar, effectively blocks any such PR until that large project is completed.

When I see #27331 I see a few new components that will benefit us, including a collapsed color swatch control. But I'm not sure we need any of the components ported over in order to get started with some of the most profound improvements, such as the progressive discovery system:

<img src="https://user-images.githubusercontent.com/548849/100453148-c578a300-30ba-11eb-8a32-e50bfc2d0888.png">

From left to right it shows how a basic default experience can be toggled to show additional tools when needed. This solves a few problems:

- we can enable design tools in the block-styling part of Global Styles, without weighing on default blocks
- we can retire every clear/reset button in favor of toggles in the ellipsis: untoggle and it also serves to reset

While the following is just a work in progress design, it shows how progressive discovery might enable a swath of rarely used design tools that are _occasionally crucial_:

<img width="1093" alt="Screenshot 2021-05-21 at 09 40 47" src="https://user-images.githubusercontent.com/1204802/119104221-4f774800-ba1c-11eb-8ec3-cd6d9f29174c.png">

Another improvement is to reduce the proliferation of inspector panels (border, color, spacing). If we carefully group design tools by topic rather than property, we can not only eliminate panels, but enhance discovery. Typography tools like font, line height and letter spacing are already grouped together. Width, height and alignment controls could live in a "Dimensions" panel. Perhaps both spacing, border controls, and any block-specific features, can live in an "Appearance" panel? Should color controls also live in these panels or does it still make sense to have a dedicated color panel?

All of these efforts will _benefit_ from some of the G2 components (such as the new stepper or segmented controls). But do we actually need them all to be merged to get started with the progressive discovery system?

More importantly, any improvements we make to inspectors controls will benefit all blocks, and specifically the progressive discovery system, more than anything, would allow us to build the tools and have them be available, while at the same time not giving them enormous prominence.

ramonjd commented:
Thanks again for your notes. I agree that designs are truly going to improve the experience and optimize the available real estate.

> I think it boils down to how much work we'll need to redo when we get a chance to revisit this. If it requires deprecations, perhaps best to wait. If not, it might be fine?

Good point. 

If we're changing the controls UI only, and not the underlying data models, then I'd expect the the block support attributes/props to remain the same and no further deprecations required. 

The block markup would also need no alternation in my opinion. 

Looking at the designs so far in https://github.com/WordPress/gutenberg/issues/27331 I can't see any radical changes to either the number or type of input controls driving block supports, so I'd be surprised if we'd need deprecations. I usually lean heavily on @aaronrobertshaw's knowledge to confirm this sort of stuff however :)

aaronrobertshaw commented:
Thanks for the help charting a path forward 👍 

> I think it boils down to how much work we'll need to redo when we get a chance to revisit this. If it requires deprecations, perhaps best to wait. If not, it might be fine?

If the underlying application of block support provided styles aren't changing and it is simply the UI evolving, I don't foresee any deprecations being required. The work will be in updating the UI.

> High level, I think the question we have to ask ourselves when we build these features is: what designs do the new tools enable, and are they worth it? While this is clearly just a basic example, it doesn't convey why we need expansive border tools urgently. It would be helpful with a pattern design to show what they can accomplish.

This is a good idea which was raised in recent discussions also. In future PRs I'll definitely look to add something illustrating the complete block pattern design to the PR description. I agree a simple screenshot or GIF illustrating the functional side of the changes doesn't given much indication as to actual need or what the feature unlocks.

In terms of this PR, I'm not sure I've seen the block patterns the feature facilitates. The most I've seen personally is that it was a part of https://github.com/WordPress/gutenberg/issues/28913. Do you have a block pattern design handy @ramonjd?

> When I see #27331 I see a few new components that will benefit us, including a collapsed color swatch control. But I'm not sure we need any of the components ported over in order to get started with some of the most profound improvements, such as the progressive discovery system:

I agree that we are not blocked in starting improvements such as the progressive discovery system. It is definitely a non-trivial change that will take a while to land.

Currently, block supports are essentially a collection of individual hooks. They are very loosely coupled so they can be used independently. In order to make a start on the new panels with the progressive discovery system, some things we'd need to address might be:

- creating new panels (complete with the menu to toggle items on/off)
- updating the logic in the primary `style.js` hook to correctly determine if a given panel should display
- updating individual border support features to implement a "reset" functionality for when toggled off
- possibly updating controls to prevent display of reset/clear buttons etc
- replicating the same changes within the Global Styles sidebar
- possibly tweaking inspector controls panel component if needed

I'm probably missing a lot as that just initial thoughts. None of that is impossible, each step will take time to go through the normal review process. 

That brings us back to finding the right balance in terms of what is genuinely needed functionality that we want to land from PRs such as this vs those that can wait. Hopefully adding pattern designs to the PRs will help in that evaluation.

ramonjd commented:
> Do you have a block pattern design handy @ramonjd?

Not handy, but I can elaborate. This PR was originally created pursuant to https://github.com/WordPress/gutenberg/issues/28913, whose objective was to **allow** block supports. 

Special cases that require complete rewiring such as pullquote, I think, weren't forseen.

For me the biggest advantage is that we're introducing flexibility. 

In the current version of this block, the color control is tightly-coupled to a custom attribute (`mainColor`), and limited by the current styles. 

This is the extent to which I can customize my pullquote block:

![May-24-2021 14-25-31](https://user-images.githubusercontent.com/6458278/119296361-1fd67300-bc9c-11eb-8281-be365f064ca8.gif)

In the current version, the styles **prescribe** when a border and background color may appear (it's one or the other), and how the text is aligned.

With block supports enabled, the supports controls better **describe** how the pullquote should look. 

Decoupling alignment from the styles alone allows me to align pullquote text to the right and left.

<img width="698" alt="Screen Shot 2021-05-24 at 2 22 05 pm" src="https://user-images.githubusercontent.com/6458278/119296452-4c8a8a80-bc9c-11eb-86c3-b1feea1907ad.png">

This issue also prosecutes the case for adding border support with examples. https://github.com/WordPress/gutenberg/issues/21540

> it doesn't convey why we need expansive border tools urgently

I don't think there is any urgency. For me it's about allowing the user more control over the look and feel of their blocks.

 If we do that, but end up making things more confusing because of suboptimal controls then I'd be supportive of deferring these code changes and related discussions if folks feel strongly about it.

Cheers!

jasmussen commented:
Thank you for the for the responses! 

> That brings us back to finding the right balance in terms of what is genuinely needed functionality that we want to land from PRs such as this vs those that can wait. Hopefully adding pattern designs to the PRs will help in that evaluation.

That is the key balance: ultimately we'll want most features available for most blocks. In absence of the progressive discovery system, we have to balance user friendliness with features: too many inspector controls come at the cost of that usability.

Starting with a mockup showing why a particular feature is needed (in this case border customization support) is an excellent way to provide context for the feature. It helps evaluate the weight of the UI added against the benefits it brings to patterns. I don't think the balance is entirely right here. 

It might be okay to land it for the Pullquote, because as noted it's a bit of an exotic block, but I also think it will be a harder sell to expand support beyond this block. Even without the progressive discovery system, we can do [improvements](https://github.com/WordPress/gutenberg/pull/30951#issuecomment-845005460) to reduce duplicate labels, and increase compression:

<img width="345" alt="Screenshot 2021-05-25 at 10 36 08" src="https://user-images.githubusercontent.com/1204802/119466623-19500600-bd45-11eb-9747-7db77d472615.png">

☝️ that on its own isn't a great solution, but it's a step in the right direction.

aaronrobertshaw commented:
> Steps towards refining the border support UI are already being made in #31585.

<img width="284" alt="Screen Shot 2021-05-26 at 9 59 52 am" src="https://user-images.githubusercontent.com/60436221/119583145-4a920a00-be09-11eb-87fb-354861b7d80a.png">

> Even without the progressive discovery system, we can do improvements to reduce duplicate labels, and increase compression:

The above screenshot is the current state of the PR previously linked [refining the border support UI](https://github.com/WordPress/gutenberg/pull/31585). It was this version of the UI I was hoping would be satisfactory enough for this Pullquote PR to be able to land.

aaronrobertshaw commented:
The refinements to the border block support UI in https://github.com/WordPress/gutenberg/pull/31585 have recently been merged. This includes all the underlying changes required such as support both shorthand and longhand styles for border radius, custom units etc. 

Work on implementing a new panel that offers progressive display options is currently under way in https://github.com/WordPress/gutenberg/pull/32392. Any block leveraging block support will pick up the new panel as that work is completed.

This PR will need another rebase to pick up the UI changes. After that, I think we might be at a point where we can complete a final review then merge, given the most immediate concerns around the UI have been resolved. 

Happy to hear from others on this?

cc @ramonjd @jasmussen

ramonjd commented:
> This PR will need another rebase to pick up the UI changes. 

Thanks for the heads up @aaronrobertshaw, and for the laying the ground work. I'll add a note to come back to this soon.

jasmussen commented:
This is what I see:

![pullquote](https://user-images.githubusercontent.com/1204802/125034051-6d624000-e090-11eb-9ae1-4aea71d71bfb.gif)

Immediate take-away, this is vastly improved, and is really fun. It's also a great choice to add it to the Pullquote block first, before expanding.

I still think there's, as we've discussed, an opportunity to refine exactly where these tools live — I could see "border" being a subset of appearance properties in a more unified panel. But there's no need for that to block this from landing, for all the reasons already stated. Nice work, thanks for your patience!

jasmussen commented:
I just noticed there are _two_ color panels in the GIF I recorded above. I can't actually reproduce how I got there, testing this PR again I'm only seeing one. Could be a temporary glitch, but thought I'd surface it.

aaronrobertshaw commented:
> I just noticed there are two color panels in the GIF I recorded above. I can't actually reproduce how I got there, testing this PR again I'm only seeing one. Could be a temporary glitch, but thought I'd surface it.

I haven't been able to reproduce the duplicate color panel either. 

Using the current state of the PR branch, rebuilding etc there were no duplicate panels in the sidebar for me. I noticed the updated border support UI in your GIF, did you rebase the PR branch at all? I'm wondering if something went awry there.

After rebasing the PR branch locally, I get the updated border support UI but still no duplicate color panel. Looking through the code itself I can't see explanation for what it's in the GIF.

On trunk, I created a post with multiple pullquote blocks using various configurations then switched to the PR branch. All the blocks migrated correctly and still no duplicate panels. 

@ramonjd Can you replicate the duplicate panel issue?

ramonjd commented:
> @ramonjd Can you replicate the duplicate panel issue?

Thanks for the ping. I've just rebased the branch and have started retesting.


https://user-images.githubusercontent.com/6458278/125233601-4622a180-e322-11eb-8c0a-a52b1663ed90.mp4


I haven't come across the double color panel conundrum yet.

aaronrobertshaw reviewed (approved):
I've given this another test after the latest rebase. Everything works as advertised for me. 

Still no sign of a duplicate color panel. I'm comfortable saying that was a one-off in Joen's local environment.

There are also a couple of e2e failures still. Re-running the tests didn't get them passing. The failures do appear unrelated to this PR so once they pass we should be right to merge. 

Thanks for all the hard work on this @ramonjd 👍

ramonjd commented:
> There are also a couple of e2e failures still. Re-running the tests didn't get them passing. The failures do appear unrelated to this PR so once they pass we should be right to merge.

Thanks a lot for the final check @aaronrobertshaw !

Looks like some of the fixtures are now failing 🤷 Maybe that's the cause of the E2E failures. I'll rebase and 🤞

ramonjd commented:
Merged changes from https://github.com/WordPress/gutenberg/pull/33502 which move block e2e fixtures

ramonjd committed:
Regenerating main color deprecation fixtures after #33502
All we needed to do was add `__deprecated` to the file name so that the fixture test regex could find it :)

ramonjd committed:
Fix after rebase
